File: backend-python/app/api/v1/user_dashboard_config.py
from typing import Optional
from fastapi import APIRouter, Depends, Query, HTTPException, status
from sqlalchemy.orm import Session
from app.database import get_db
from app.services.user_dashboard_config import UserDashboardConfigService
from app.schemas.common import ApiResponse


router = APIRouter(prefix="/user-dashboard-config", tags=["User Dashboard Config"])


@router.get("/{dashboard_type}", response_model=ApiResponse)
def get_dashboard_config(
    dashboard_type: str,
    user_id: Optional[str] = Query(None, description="User ID"),
    db: Session = Depends(get_db)
):
    service = UserDashboardConfigService(db)
    user_id = user_id or 'default'
    config = service.get_by_user_and_type(user_id, dashboard_type)
    return ApiResponse(
        code=200,
        message="success",
        data=config.to_dict()
    )


@router.post("", response_model=ApiResponse)
def save_dashboard_config(
    user_id: str = Query(..., description="User ID"),
    dashboard_type: str = Query(..., description="Dashboard type"),
    config: dict = None,
    db: Session = Depends(get_db)
):
    service = UserDashboardConfigService(db)
    config = service.save_config(user_id, dashboard_type, config)
    return ApiResponse(
        code=200,
        message="Saved successfully",
        data=config.to_dict()
    )


@router.delete("/{dashboard_type}", response_model=ApiResponse)
def delete_dashboard_config(
    dashboard_type: str,
    user_id: str = Query(..., description="User ID"),
    db: Session = Depends(get_db)
):
    service = UserDashboardConfigService(db)
    service.delete_config(user_id, dashboard_type)
    return ApiResponse(
        code=200,
        message="Deleted successfully",
        data=None
    )


@router.get("/all/{user_id}", response_model=ApiResponse)
def get_all_dashboard_configs(
    user_id: str,
    db: Session = Depends(get_db)
):
    service = UserDashboardConfigService(db)
    configs = service.get_all_by_user(user_id)
    return ApiResponse(
        code=200,
        message="success",
        data=[config.to_dict() for config in configs]
    )

File: src/services/userDashboardConfig.ts
import apiClient from '../utils/api'

export interface DashboardConfig {
  cards: Array<{
    id: string
    visible: boolean
    position: number
  }>
  charts: Array<{
    id: string
    visible: boolean
    position: number
  }>
  layout: 'grid' | 'list'
}

export interface ApiResponse<T = any> {
  code: number
  message: string
  data: T
}

export const userDashboardConfigService = {
  async getConfig(dashboardType: string, userId?: string): Promise<ApiResponse<DashboardConfig>> {
    const params = userId ? { user_id: userId } : {}
    return await apiClient.get(`/user-dashboard-config/${dashboardType}`, { params })
  },

  async saveConfig(dashboardType: string, config: DashboardConfig, userId?: string): Promise<ApiResponse<DashboardConfig>> {
    const params = userId ? { user_id: userId, dashboard_type: dashboardType } : { dashboard_type: dashboardType }
    return await apiClient.post('/user-dashboard-config', config, { params })
  },

  async deleteConfig(dashboardType: string, userId?: string): Promise<ApiResponse<null>> {
    const params = userId ? { user_id: userId, dashboard_type: dashboardType } : { dashboard_type: dashboardType }
    return await apiClient.delete(`/user-dashboard-config/${dashboardType}`, { params })
  },

  async getAllConfigs(userId: string): Promise<ApiResponse<DashboardConfig[]>> {
    return await apiClient.get(`/user-dashboard-config/all/${userId}`)
  }
}
